service/frontend/handler.go

```diff
@@ -1619,54 +1582,117 @@ func createDomainResponse(info *persistence.DomainInfo, config *persistence.Doma
 	return i, c
 }
 
-func (wh *WorkflowHandler) createPollForDecisionTaskResponse(ctx context.Context,
-	matchingResponse *m.PollForDecisionTaskResponse, history *gen.History, nextPageToken []byte) *gen.PollForDecisionTaskResponse {
-	resp := &gen.PollForDecisionTaskResponse{}
-	if matchingResponse != nil {
-		resp.TaskToken = matchingResponse.TaskToken
-		resp.WorkflowExecution = matchingResponse.WorkflowExecution
-		resp.WorkflowType = matchingResponse.WorkflowType
-		resp.PreviousStartedEventId = matchingResponse.PreviousStartedEventId
-		resp.StartedEventId = matchingResponse.StartedEventId
-		resp.Query = matchingResponse.Query
-		resp.BacklogCountHint = matchingResponse.BacklogCountHint
-	}
-
-	if matchingResponse.Query != nil && resp.WorkflowType == nil {
-		// for query task, the matching engine was not able to populate the WorkflowType, so set here from history
-		if history != nil && len(history.Events) > 0 &&
-			history.Events[0].WorkflowExecutionStartedEventAttributes != nil {
-			resp.WorkflowType = history.Events[0].WorkflowExecutionStartedEventAttributes.WorkflowType
-		} else {
+func (wh *WorkflowHandler) createPollForDecisionTaskResponse(ctx context.Context, domainID string,
+	matchingResp *m.PollForDecisionTaskResponse) (*gen.PollForDecisionTaskResponse, error) {
+
+	if matchingResp.WorkflowExecution == nil {
+		// this will happen if there is no decision task to be send to worker / caller
+		return &gen.PollForDecisionTaskResponse{}, nil
+	}
+
+	var history *gen.History
+	var continuation []byte
+	var err error
+
+	// if matchingResp.WorkflowExecution.RunId == nil {
+	// 	wh.Service.GetLogger().Errorf(
+	// 		"PollForDecisionTask from matching engine doesn't have run id. TaskList: %v",
+	// 		*pollRequest.TaskList.Name)
+	// 	return nil, wh.error(errRunIDNotSet, scope)
+	// }
+
+	fmt.Println("@@@@@@@@@@@@@@")
+	fmt.Println(matchingResp)
+	fmt.Println(matchingResp.GetStickyExecutionEnabled())
+	fmt.Println(matchingResp.Query)
+	fmt.Println("@@@@@@@@@@@@@@")
+	if matchingResp.GetStickyExecutionEnabled() && matchingResp.Query != nil {
+		// meaning sticky query, we should not return any events to worker
```

Comment:
your client side change will need to handle the case where the query task comes in with no history, but the cache is already evicted, in that case you have to reload the history from client side, but need to avoid putting that workflowExecutionContext into cache.

Reply:
i believe i have already handled that case:
handle of cache miss:
https://github.com/uber-go/cadence-client/blob/844db2a5659db29680c1bb8ea458b3dc3cc40f49/internal/internal_task_handlers.go#L496

avoiding cache:
https://github.com/uber-go/cadence-client/blob/844db2a5659db29680c1bb8ea458b3dc3cc40f49/internal/internal_task_handlers.go#L509